File: application/backend-credit-card-enrollment/backend-java/src/main/java/cloud/ambar/common/projection/ProjectionController.java
package cloud.ambar.common.projection;

import cloud.ambar.common.ambar.AmbarHttpRequest;
import cloud.ambar.common.ambar.AmbarResponseFactory;
import cloud.ambar.common.event.Event;
import cloud.ambar.common.serializedevent.Deserializer;
import lombok.*;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.data.mongodb.core.query.Query;

import java.util.Arrays;
import java.util.stream.Collectors;

@RequiredArgsConstructor
public abstract class ProjectionController {
    private final MongoTransactionalProjectionOperator mongoTransactionalProjectionOperator;
    private final Deserializer deserializer;
    private static final Logger log = LogManager.getLogger(ProjectionController.class);

    protected String processProjectionHttpRequest(
            final AmbarHttpRequest ambarHttpRequest,
            final ProjectionHandler projectionHandler,
            final String projectionName) {
        log.info("Reaction controller received http request: " + ambarHttpRequest);

        try {
            Event event = deserializer.deserialize(ambarHttpRequest.getSerializedEvent());

            // We start a Mongo transaction because if a projection handler needs to update a projection,
            // it should do so idempotently by checking if the event has already been projected,
            // and it should do so with a transaction to not receive dirty reads.
            mongoTransactionalProjectionOperator.startTransaction();
            boolean isAlreadyProjected = mongoTransactionalProjectionOperator.operate().count(
                Query.query(
                        Criteria.where("eventId").is(event.getEventId())
                                .and("projectionName").is(projectionName)
                ),
                "ProjectionIdempotency_ProjectedEvent"
            ) != 0;

            if (isAlreadyProjected) {
                mongoTransactionalProjectionOperator.abortDanglingTransactionsAndReturnSessionToPool();
                return AmbarResponseFactory.successResponse();
            }

            mongoTransactionalProjectionOperator.operate().save(
                    ProjectedEvent.builder()
                            .eventId(event.getEventId())
                            .projectionName(projectionName)
                            .build(),
                    "ProjectionIdempotency_ProjectedEvent"
            );

            projectionHandler.project(event);

            mongoTransactionalProjectionOperator.commitTransaction();
            mongoTransactionalProjectionOperator.abortDanglingTransactionsAndReturnSessionToPool();

            return AmbarResponseFactory.successResponse();
        } catch (Exception e) {
            if (e.getMessage() != null && e.getMessage().startsWith("Unknown event type")) {
                log.warn("Unknown event type. Skipping projection.");
                log.warn(e);

                mongoTransactionalProjectionOperator.abortDanglingTransactionsAndReturnSessionToPool();

                return AmbarResponseFactory.successResponse();
            }

            log.error("Failed to process projection http request.");
            log.error(e);
            log.error(e.getMessage());
            String stackTraceString = Arrays.stream(e.getStackTrace())
                    .map(StackTraceElement::toString)
                    .collect(Collectors.joining("\n"));
            log.error(stackTraceString);

            mongoTransactionalProjectionOperator.abortDanglingTransactionsAndReturnSessionToPool();

            return AmbarResponseFactory.retryResponse(e);
        }
    }

    @Builder
    @Setter
    @Getter
    public static class ProjectedEvent {
        @NonNull
        private String eventId;
        @NonNull private String projectionName;
    }
}

File: application/backend-credit-card-enrollment/backend-c#/CreditCardEnrollment/Common/Projection/ProjectionController.cs
using CreditCardEnrollment.Common.Ambar;
using CreditCardEnrollment.Common.Events;
using Microsoft.AspNetCore.Mvc;
using MongoDB.Driver;
using System.Text.Json;

namespace CreditCardEnrollment.Common.Projection;

public abstract class ProjectionController : ControllerBase
{
    private readonly IMongoTransactionalProjectionOperator _mongoOperator;
    private readonly IDeserializer _deserializer;
    private readonly IMongoDatabase _database;
    private readonly ILogger<ProjectionController> _logger;

    protected ProjectionController(
        IMongoTransactionalProjectionOperator mongoOperator,
        IDeserializer deserializer,
        IMongoDatabase database,
        ILogger<ProjectionController> logger)
    {
        _mongoOperator = mongoOperator;
        _deserializer = deserializer;
        _database = database;
        _logger = logger;
    }

    protected async Task<IActionResult> ProcessProjectionHttpRequest(AmbarHttpRequest request, ProjectionHandler handler, string projectionName)
    {
        _logger.LogInformation(
            "Processing projection request. RequestId: {RequestId}, Projection: {ProjectionName}, EventId: {EventId}, EventName: {EventName}",
            HttpContext.TraceIdentifier,
            projectionName,
            request.SerializedEvent.EventId,
            request.SerializedEvent.EventName
        );

        try
        {
            var @event = _deserializer.Deserialize(request.SerializedEvent.JsonPayload);

            await _mongoOperator.ExecuteInTransaction(async () =>
            {
                _logger.LogInformation(
                    "Started MongoDB transaction for projection {ProjectionName}, EventId: {EventId}",
                    projectionName,
                    request.SerializedEvent.EventId
                );

                var collection = _database.GetCollection<ProjectedEvent>("ProjectionIdempotency_ProjectedEvent");
                var filter = Builders<ProjectedEvent>.Filter.And(
                    Builders<ProjectedEvent>.Filter.Eq(p => p.EventId, @event.EventId),
                    Builders<ProjectedEvent>.Filter.Eq(p => p.ProjectionName, projectionName)
                );

                var isAlreadyProjected = await collection.Find(filter).AnyAsync();
                if (isAlreadyProjected)
                {
                    _logger.LogInformation(
                        "Event {EventId} already projected for {ProjectionName}",
                        @event.EventId,
                        projectionName
                    );
                    return;
                }

                await collection.InsertOneAsync(new ProjectedEvent(@event.EventId, projectionName));
                
                try
                {
                    handler.HandleEvent(@event);
                    _logger.LogInformation(
                        "Successfully processed projection {ProjectionName} for event {EventId}",
                        projectionName,
                        @event.EventId
                    );
                }
                catch (Exception ex)
                {
                    _logger.LogError(
                        ex,
                        "Error during projection processing for {ProjectionName}, EventId: {EventId}",
                        projectionName,
                        @event.EventId
                    );
                    throw;
                }
            });

            _logger.LogInformation(
                "Successfully completed projection {ProjectionName} for event {EventId}",
                projectionName,
                @event.EventId
            );
            return Ok();
        }
        catch (Exception ex)
        {
            if (ex is ArgumentException argEx && argEx.Message.StartsWith("Unknown event type"))
            {
                _logger.LogWarning(
                    ex,
                    "Unknown event type in request. Projection: {ProjectionName}, EventId: {EventId}, EventName: {EventName}",
                    projectionName,
                    request.SerializedEvent.EventId,
                    request.SerializedEvent.EventName
                );
                return Ok();
            }

            _logger.LogError(
                ex,
                "Failed to process projection request. Projection: {ProjectionName}, EventId: {EventId}, EventName: {EventName}, JsonPayload: {JsonPayload}",
                projectionName,
                request.SerializedEvent.EventId,
                request.SerializedEvent.EventName,
                request.SerializedEvent.JsonPayload
            );
            return StatusCode(500, new
            {
                error = "Failed to process projection",
                message = ex.Message,
                details = ex.ToString()
            });
        }
    }
}
